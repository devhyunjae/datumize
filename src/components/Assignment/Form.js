import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Card, Button, Toaster, Position } from '@blueprintjs/core'

import UserDropDown from './UsersDropdown'
import ProjectsDropDown from './ProjectsDropdown'
import RolesDropDown from './RolesDropdown'
import * as actions from '../../actions/users'

const Container = styled.div`
  display: flex;
  flex: 1;
  justify-content: center;
  align-items: flex-start;
  flex-direction: column;
`

class Form extends PureComponent {
  static propTypes = {
    updateUser: PropTypes.func.isRequired,
    successMessage: PropTypes.string,
  }

  static defaultProps = {
    successMessage: null,
  }

  state = {
    selectedUser: null,
    selectedProject: null,
    selectedRole: null,
  }

  componentDidUpdate(prevProps) {
    const { successMessage } = this.props

    if (successMessage && (prevProps.successMessage !== successMessage)) {
      this.toaster.show({ message: successMessage })
    }
  }

  onClick = () => {
    const { selectedUser, selectedProject, selectedRole } = this.state
    this.props.updateUser({ user: selectedUser, project: selectedProject, role: selectedRole })
  }

  setSelectedUser = (selectedUser) => {
    this.setState({
      selectedUser,
    })
  }

  setSelectedProject = (selectedProject) => {
    this.setState({
      selectedProject,
    })
  }

  setSelectedRole = (selectedRole) => {
    this.setState({
      selectedRole,
    })
  }

  toasterHandler = (ref) => {
    this.toaster = ref
  }

  render() {
    return (
      <Card elevation={Card.ELEVATION_TWO}>
        <Container>
          <UserDropDown setSelectedUser={this.setSelectedUser} />
          <ProjectsDropDown setSelectedProject={this.setSelectedProject} />
          <RolesDropDown setSelectedRole={this.setSelectedRole} />
          <Button type="button" onClick={this.onClick} className="pt-fill">
            Save
          </Button>
        </Container>
        <Toaster position={Position.TOP} ref={this.toasterHandler} />
      </Card>
    )
  }
}

const mapStateToProps = state => ({
  successMessage: state.users.success,
})

export default connect(
  mapStateToProps,
  actions,
)(Form)
