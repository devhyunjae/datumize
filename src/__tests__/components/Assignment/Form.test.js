import React from 'react'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'

import Form from '../../../components/Assignment/Form'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)
const store = mockStore(
  {
    users: {
      users: [{
        id: 1, name: 'John Doe',
      }],
    },
    projects: {
      projects: [{
        id: 1, name: 'Surf House in Canary Island',
      }],
    },
    roles: {
      roles: [{
        id: 1, name: 'CEO',
      }],
    },
  },
)

describe('<Form />', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Form />
      </Provider>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
