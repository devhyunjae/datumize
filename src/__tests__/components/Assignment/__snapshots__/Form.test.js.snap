// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> renders correctly 1`] = `
<div
  className="pt-card pt-elevation-0"
  onClick={undefined}
>
  <div
    className="sc-bwzfXH gNjnI"
  >
    <div
      className="sc-bdVaJa ePGVPP"
    >
      <h5>
        1. Choose a user
      </h5>
      <div
        className="pt-select pt-fill"
      >
        <select
          onChange={[Function]}
        >
          <option
            name="John Doe"
            value={1}
          >
            John Doe
          </option>
        </select>
      </div>
    </div>
    <div
      className="sc-bdVaJa ePGVPP"
    >
      <h5>
        2. Choose a project
      </h5>
      <div
        className="pt-select pt-fill"
      >
        <select
          onChange={[Function]}
        >
          <option
            value={1}
          >
            Surf House in Canary Island
          </option>
        </select>
      </div>
    </div>
    <div
      className="sc-bdVaJa ePGVPP"
    >
      <h5>
        3. Choose a role
      </h5>
      <div
        className="pt-select pt-fill"
      >
        <select
          onChange={[Function]}
        >
          <option
            value={1}
          >
            CEO
          </option>
        </select>
      </div>
    </div>
    <button
      className="pt-button pt-fill"
      disabled={undefined}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span>
        Save
      </span>
    </button>
  </div>
</div>
`;
